helper_crates/vtable/src/vrc.rs

```diff
@@ -0,0 +1,206 @@
+        Self { size: layout.size(), align: layout.align() }
+    }
+}
+
+impl core::convert::TryFrom<Layout> for core::alloc::Layout {
+    type Error = core::alloc::LayoutErr;
+
+    fn try_from(value: Layout) -> Result<Self, Self::Error> {
+        Self::from_size_align(value.size, value.align)
+    }
+}
+
+#[repr(C)]
+struct VRcInner<'vt, T: VTableMeta, X> {
+    vtable: &'vt T::VTable,
+    strong_ref: Cell<u32>,
+    weak_ref: Cell<u32>,
+    data: X,
+}
+
+/// A reference counted pointer to an object matching the virtual table `T`
+///
+/// Similar to [`std::rc::Rc`] where T is a VTable struct. But it has the following difference with
+/// the type from std:
+/// - It doesn't implement `Deref`, but you need to call `borrow()` to access a `VRef<R>`
+/// which implements `Deref`.
+/// - It does not allow to access mutable reference. (No `get_mut` or `make_mut`), meaning it is
+/// safe to get a Pin reference with `borrow_pin`.
+/// - It is safe to pass it accross ffi boundaries.
+#[repr(transparent)]
+pub struct VRc<T: VTableMetaDropInPlace + 'static> {
+    inner: NonNull<VRcInner<'static, T, u8>>,
+}
+
+impl<T: VTableMetaDropInPlace + 'static> Drop for VRc<T> {
+    fn drop(&mut self) {
+        unsafe {
+            let inner = self.inner.as_ref();
+            inner.strong_ref.set(inner.strong_ref.get() - 1);
+            if inner.strong_ref.get() == 0 {
+                let data = &inner.data as *const u8 as *mut u8;
+                let layout = T::drop_in_place(inner.vtable, data);
+                inner.weak_ref.set(inner.weak_ref.get() - 1);
+                if inner.weak_ref.get() == 0 {
+                    let vtable = inner.vtable;
+                    T::dealloc(vtable, self.inner.cast().as_ptr(), layout);
+                } else {
+                    self.inner.cast::<VRcInner<T, Layout>>().as_mut().data = layout;
+                }
+            }
+        }
+    }
+}
+
+impl<T: VTableMetaDropInPlace> VRc<T> {
+    /// Create a new VRc from an instance of a type that can be associated with a VTable.
+    ///
+    /// Will move the instance on the heap.
+    ///
+    /// (the `HasStaticVTable` is implemented by the `“MyTrait”VTable_static!` macro generated by
+    /// the #[vtable] macro)
+    pub fn new<X: HasStaticVTable<T>>(data: X) -> Self {
+        let layout = core::alloc::Layout::new::<VRcInner<T, X>>();
+        let other = core::alloc::Layout::new::<VRcInner<T, Layout>>();
+        let layout = core::alloc::Layout::from_size_align(
+            layout.size().max(other.size()),
+            layout.align().max(other.align()),
+        )
+        .unwrap();
+        let mem = unsafe { std::alloc::alloc(layout) as *mut VRcInner<T, X> };
+        let inner = NonNull::new(mem).unwrap();
+        assert!(!mem.is_null());
+
+        unsafe {
+            mem.write(VRcInner {
+                vtable: X::static_vtable(),
+                strong_ref: Cell::new(1),
+                weak_ref: Cell::new(1),
```

Comment:
Shouldn't the initial weak count be 0?

Reply:
1 is correct, because VRc::drop is decreasing the weak_ref count when the strong_count reaches 0.
Added comments